test/MoveAdd.lean

```diff
@@ -0,0 +1,98 @@
+import Mathlib.Tactic.MoveAdd
+import Mathlib.Data.Nat.Basic
+
+universe u
+
+section tactic
+open Mathlib.MoveAdd
+
+run_cmd if uniquify [ 0,      1,      0,      1,      0,      3] ≠
+                    [(0, 0), (1, 0), (0, 1), (1, 1), (0, 2), (3, 0)] then
+          throwError "'uniquify' is not uniquifying"
+
+run_cmd
+  let dat := [(0, true), (1, false), (2, true)]
+  if (#[0, 1, 2, 3, 4].qsort (fun x y => (weight dat x) ≤ (weight dat y)) != #[0, 2, 3, 4, 1]) then
+    throwError "The sorting order induced by 'wtFnc' does not seem to have the required properties"
+
+run_cmd if false = (reorderUsing [0, 1, 2] [(0, false)] = [1, 2, 0] &&
+                    reorderUsing [0, 1, 2] [(1, true)] = [1, 0, 2] &&
+                    reorderUsing [0, 1, 2] [(1, true), (0, false)] = [1, 2, 0]) then
+          throwError "'reorderUsing' is not reordering properly"
+
+run_cmd if reorderUsing [1, 5, 4, 3, 2, 1] [(3, true), (2, false), (1, false)] ≠
+                        [3, 5, 4, 1, 2, 1] then
+          throwError "'reorderUsing' is not reordering properly"
+
+end tactic
+
+variable {R : Type u}
+section add
+
+section semigroup
+variable [AddCommSemigroup R] {a b c d e f g h : R}
+
+example (h : a + b + c = d) : b + (a + c) = d := by
+  move_add [← a]
+  assumption
+
+example : let k := c + (a + b); k = a + b + c := by
+  move_add [← a, c]
+  rfl
+
+example (h : b + a = b + c + a) : a + b = a + b + c :=
+by move_add [a]; assumption
+
+example [Mul R] [Neg R] : a + (b + c + a) * (- (d + e) + e) + f + g =
+    (c + b + a) * (e + - (e + d)) + g + f + a := by
+  move_add [b, d, g, f, a, e]; rfl
```

Comment:
To be honest, I find this very confusing, and have no idea what the intended behaviour of this tactic is! How many different lists are being reordered here??

Reply:
I added a more detailed description of the algorithm at the beginning of the other file.  While doing so, I realized how many implicit conventions I was using: I hope that I made enough of them explicit!

Reply:
To answer your question, the `move_add` call in this example places (any subset of) its inputs to the right of each sum in which they appear, preserving the calling order.  Thus, `e` will be the rightmost summand of each sum where it appears, the next rightmost term will be `a`, then `f`, and so on.